src/ua_types.c

```diff
@@ -114,6 +115,43 @@ UA_cleanupDataTypeWithCustom(const UA_DataTypeArray *customTypes) {
 /*****************/
 /* Builtin Types */
 /*****************/
+#define FORMAT_STACK_SIZE 256
+
+static UA_String
+vfromFormatWithBuffer (char *buffer, size_t size, char *format, va_list args)
+{
+    if(!buffer || size < 1) return UA_STRING_NULL;
+    UA_String s;
+    s.length = mp_vsnprintf (buffer, size, format, args);
+    s.length = UA_MIN(s.length, size);
+    s.data = (UA_Byte*) buffer;
+    return s;
+}
+
+UA_String
+UA_String_fromFormatWithBuffer (char *buffer, size_t size, char *format, ...)
+{
+    if(!buffer || size < 1) return UA_STRING_NULL;
+    va_list args;
+    va_start (args, format);
+    UA_String s = vfromFormatWithBuffer (buffer, size, format, args);
+    va_end(args);
+    return s;
+}
+
+UA_String
+UA_String_fromFormat (char *format, ...)
+{
+    char tmp[FORMAT_STACK_SIZE];
+    va_list args;
+    va_start (args, format);
+    UA_String s = vfromFormatWithBuffer (tmp, FORMAT_STACK_SIZE, format,args);
```

Comment:
Check whether the string length was enough.
If not, take the required size (returned from printf), malloc and print again.

Only do the memcpy if printing into the buffer succeeded.